pkg/networkservice/chains/chainstest/builder.go

```diff
@@ -0,0 +1,264 @@
+	return b
+}
+
+// SetRegistrySupplier replaces default memory registry supplier to custom function
+func (b *DomainBuilder) SetRegistrySupplier(f SupplyRegistryFunc) *DomainBuilder {
+	b.supplyRegistry = f
+	return b
+}
+
+// SetForwarderSupplier replaces default dummy forwarder supplier to custom function
+func (b *DomainBuilder) SetForwarderSupplier(f SupplyForwarderFunc) *DomainBuilder {
+	b.supplyForwarder = f
+	return b
+}
+
+// SetNSMgrSupplier replaces default nsmgr supplier to custom function
+func (b *DomainBuilder) SetNSMgrSupplier(f SupplyNSMgrFunc) *DomainBuilder {
+	b.supplyNSMgr = f
+	return b
+}
+
+func (b *DomainBuilder) dialContext(ctx context.Context, u *url.URL) *grpc.ClientConn {
+	conn, err := grpc.DialContext(ctx, grpcutils.URLToTarget(u),
+		grpc.WithInsecure(),
+		grpc.WithBlock(),
+		grpc.WithDefaultCallOptions(grpc.WaitForReady(true)))
+	b.resources = append(b.resources, func() {
+		_ = conn.Close()
+	})
+	b.require.NoError(err, "Can not dial to", u)
+	return conn
+}
+
+func (b *DomainBuilder) newNSMgr(ctx context.Context, registryURL *url.URL) (nsmgr.Nsmgr, *url.URL) {
+	var registryCC *grpc.ClientConn
+	if registryURL != nil {
+		registryCC = b.dialContext(ctx, registryURL)
+	}
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
+	b.require.NoError(err)
+	serveURL := &url.URL{Scheme: "tcp", Host: "127.0.0.1:" + fmt.Sprint(listener.Addr().(*net.TCPAddr).Port)}
+	b.require.NoError(listener.Close())
+
+	nsmgrReg := &registryapi.NetworkServiceEndpoint{
+		Name: "nsmgr-" + uuid.New().String(),
+		Url:  serveURL.String(),
+	}
+
+	mgr := b.supplyNSMgr(ctx, nsmgrReg, authorize.NewServer(), b.generateTokenFunc, registryCC, grpc.WithInsecure(), grpc.WithBlock(), grpc.WithDefaultCallOptions(grpc.WaitForReady(true)))
+
+	serve(ctx, serveURL, mgr.Register)
+	log.Entry(ctx).Infof("NSMgr %v listen on: %v", nsmgrReg.Name, serveURL)
+	return mgr, serveURL
+}
+
+func serve(ctx context.Context, u *url.URL, register func(server *grpc.Server)) {
+	server := grpc.NewServer()
+	register(server)
+	errCh := grpcutils.ListenAndServe(ctx, u, server)
+	go func() {
+		select {
+		case <-ctx.Done():
+			log.Entry(ctx).Infof("Stop serve: %v", u.String())
+			return
+		case err := <-errCh:
+			log.Entry(ctx).Fatalf("An error during serve: %v", err.Error())
+		}
+	}()
+}
+func (b *DomainBuilder) newCrossConnectNSE(ctx context.Context, name string, connectTo *url.URL) (endpoint.Endpoint, *url.URL) {
+	crossNSE := b.supplyForwarder(ctx, name, b.generateTokenFunc, connectTo, grpc.WithInsecure(), grpc.WithDefaultCallOptions(grpc.WaitForReady(true)))
+	serveURL := &url.URL{Scheme: "tcp", Host: "127.0.0.1:0"}
+	serve(ctx, serveURL, crossNSE.Register)
+	log.Entry(ctx).Infof("%v listen on: %v", name, serveURL)
+	return crossNSE, serveURL
+}
+
+func (b *DomainBuilder) newRegistry(ctx context.Context, supply SupplyRegistryFunc) (registry.Registry, *url.URL) {
+	result := supply()
+	serveURL := &url.URL{Scheme: "tcp", Host: "127.0.0.1:0"}
```

Comment:
Same.

Reply:
Let's discuss this moment here #460 (comment)